### PR TITLE
Support 'latest' version, explicit binary URL for Embedded DB

### PR DESCRIPTION
**Pull request by parkerduckworth**
- Allow `'latest'` version to automatically download the latest Weaviate server version
- Support explicit URL to weaviate binary in the form of `binaryUrl` property of `EmbeddedOptions`
- Detect machine architecture to download correct binary

### CODE REVIEW

File: src/embedded/index.ts [hunk 1]
```diff
@@ -67,11 +67,15 @@ export class EmbeddedOptions {
...
+      return 'latest';
+    }
     if (cfg.version.match(/[1-9]\.[1-9]{2}\..*/g)) {
       return cfg.version;
-    } else {
-      throw new Error(`invalid version: ${cfg.version}. version must resemble '{major}.{minor}.{patch}'`);
     }
+    throw new Error(
+      `invalid version: ${cfg.version}. version must resemble '{major}.{minor}.{patch}, or 'latest'`
+    );
```
Comment (dirkkul) [inline src/embedded/index.ts:78]: I think we still should support URLs. In the future we might build binaries for each PR

File: src/embedded/index.ts [hunk 2]
```diff
@@ -145,80 +151,173 @@ export class EmbeddedDB {
...
-        strict: true,
-      })
-        .on('finish', () => {
-          tarball.close();
-          fs.renameSync(join(dirname(opt.binaryPath), 'weaviate'), opt.binaryPath);
-          resolve(null);
+  private buildBinaryUrl(): string {
+    let arch: string;
+    switch (process.arch) {
+      case 'arm64':
```
Comment (dirkkul) [inline src/embedded/index.ts:242]: did you test this?
On python the platform command returns aarch64 (which is taken from uname), no idea what the TS library does

(Afaik we do not have anyone with an ARM64 laptop)

File: src/embedded/index.ts [hunk 3]
```diff
@@ -145,80 +151,173 @@ export class EmbeddedDB {
...
+    return new Promise((resolve, reject) => {
+      tarball.pipe(
+        extract({
+          cwd: dirname(tarballPath),
+          strict: true,
         })
-        .on('error', function (err) {
-          reject(new Error(`failed to untar binary: ${JSON.stringify(err)}`));
+          .on('finish', () => {
+            tarball.close();
```
Comment (dirkkul) [inline src/embedded/index.ts:267]: delete tarbal after unpacking? :)

File: src/embedded/index.ts [hunk 4]
```diff
@@ -145,80 +151,173 @@ export class EmbeddedDB {
...
+            resolve(null);
+          })
+          .on('error', (err) => {
+            reject(new Error(`failed to untar binary: ${JSON.stringify(err)}`));
+          })
+      );
+    });
+  }
+
+  private waitTillListening(): Promise<null> {
```
Comment (dirkkul) [inline src/embedded/index.ts:278]: just a note, this function could be used to implement [this here](https://github.com/weaviate/typescript-client/issues/17)
(obviously should not be part of this PR)

File: src/embedded/index.ts (hunk 2 above)
Comment (parkerduckworth) [reply src/embedded/index.ts:242]: Yes, I have tested on both arm64 and amd64 environments.

Also see: https://nodejs.org/api/process.html#processarch

File: src/embedded/index.ts (hunk 1 above)
Comment (parkerduckworth) [reply src/embedded/index.ts:89]: done
